<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>CIA BENCHMARKING &mdash; CIA 0.0.post1.dev34+gb32d426.d20220708 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="License" href="../../license.html" />
    <link rel="prev" title="CLUSTER INDEPENDENT ANNOTATION" href="../workflow/Cluster_Independent_Annotation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> CIA
          </a>
              <div class="version">
                0.0.post1.dev34+gb32d426.d20220708
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules_modified.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workflow/Cluster_Independent_Annotation.html">Tutorial: CIA workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tutorial: CIA benchmarking</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#description">Description</a></li>
<li class="toctree-l2"><a class="reference internal" href="#training-dataset">Training dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#test-dataset">Test dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="#cluster-based-annotation">Cluster-based annotation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#scorect">Scorect</a></li>
<li class="toctree-l3"><a class="reference internal" href="#performances">Performances</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#cell-level-classification">Cell level classification</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#celltypist">Celltypist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#besca">Besca</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">Performances</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#over-clustering-driven-approach">Over-clustering driven approach</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#celltypist-majority-voting">Celltypist majority voting</a></li>
<li class="toctree-l3"><a class="reference internal" href="#majority-voting-embedded-in-cia">Majority voting embedded in CIA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#majority-voting-applied-to-besca">Majority voting applied to Besca</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2">Performances</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#conclusions">Conclusions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../changelog.html">Changelog</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">CIA</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
      <li>CIA BENCHMARKING</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/tutorial/benchmark/CIA_benchmarking.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="cia-benchmarking">
<h1>CIA BENCHMARKING<a class="headerlink" href="#cia-benchmarking" title="Permalink to this headline">ÔÉÅ</a></h1>
<div class="section" id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this headline">ÔÉÅ</a></h2>
<p>In this notebook we compared <strong>CIA</strong> with <strong>other classification tools</strong>
among the newest and all compatible with Scanpy workflow: - <strong>Scorect</strong>
(<a class="reference external" href="https://github.com/LucasESBS/scoreCT">Github public repository</a>),
which requires a gmt file with signatures and a test dataset already
divided into clusters with Differentially Expressed Genes (DEGs) for
each of them. It relies on the division of <strong>ranked DEGs into bins</strong> and
the calculation of a <strong>weighted score</strong> for each cluster depending on
how many DEGs of each bin are also present in the gmt signatures. -
<strong>Celltypist</strong> (Dom√≠nguez et al., Science, 2022
<a class="reference external" href="https://www.science.org/doi/10.1126/science.abl5197">[1]</a>) which,
exploiting a <strong>combined approach based on Logistic Regression and
Stocastic Gradient Descent</strong>, extracts the most revelevant feature of
<strong>each cluster of the training set</strong> and <strong>trains a classification
model</strong>. - <strong>Besca</strong> auto_annot module (M√§dler et al., NAR Genomics and
Bioinformatics, 2021 <a class="reference external" href="https://doi.org/10.1093/nargab/lqab102">[2]</a>),
which can be run with <strong>Logistic Regression</strong> (LR) or <strong>Support Vector
Machine</strong> (SVM) algorithms.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">partial</span>
<span class="kn">from</span> <span class="nn">scipy.sparse</span> <span class="kn">import</span> <span class="n">issparse</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">sklearn</span> <span class="kn">import</span> <span class="n">metrics</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span>
<span class="kn">from</span> <span class="nn">cia</span> <span class="kn">import</span> <span class="n">investigate</span><span class="p">,</span> <span class="n">report</span><span class="p">,</span> <span class="n">external</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">celltypist</span>
<span class="kn">import</span> <span class="nn">scorect</span> <span class="k">as</span> <span class="nn">ct</span>
<span class="kn">import</span> <span class="nn">besca</span> <span class="k">as</span> <span class="nn">bc</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Global</span> <span class="n">seed</span> <span class="nb">set</span> <span class="n">to</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
<div class="section" id="training-dataset">
<h2>Training dataset<a class="headerlink" href="#training-dataset" title="Permalink to this headline">ÔÉÅ</a></h2>
<p>We <strong>trained all the classifiers</strong> with Hao et al 2021
<a class="reference external" href="https://www.sciencedirect.com/science/article/pii/S0092867421005833">[3]</a>
<strong>PBMC atlas</strong>, the same from which we extracted the gene signatures in
<a class="reference external" href="../workflow/Cluster_Independent_Annotation.ipynb">CIA workflow
tutorial</a>.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>cp ./data/atlas_final.h5ad.gz  ./data/copy_atlas.h5ad.gz
<span class="o">!</span>gunzip ./data/copy_atlas.h5ad.gz
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># to read the atlas data</span>
<span class="n">adata</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;./data/copy_atlas.h5ad&#39;</span><span class="p">)</span>
<span class="n">adata</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs √ó n_vars = 154975 √ó 1555
    obs: &#39;nCount_ADT&#39;, &#39;nFeature_ADT&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;orig.ident&#39;, &#39;lane&#39;, &#39;donor&#39;, &#39;time&#39;, &#39;celltype.l1&#39;, &#39;celltype.l2&#39;, &#39;celltype.l3&#39;, &#39;Phase&#39;, &#39;nCount_SCT&#39;, &#39;nFeature_SCT&#39;, &#39;Cell type&#39;, &#39;B&#39;, &#39;CD4 T&#39;, &#39;CD8 T&#39;, &#39;DC&#39;, &#39;Mono&#39;, &#39;NK&#39;, &#39;Platelet&#39;, &#39;CD8 T_negative&#39;, &#39;CD4 T_negative&#39;, &#39;Prediction fast mode&#39;, &#39;Prediction standard mode&#39;, &#39;Prediction q&#39;, &#39;Prediction p-val&#39;, &#39;B_filtered_FC&#39;, &#39;CD4 T_filtered_FC&#39;, &#39;CD8 T_filtered_FC&#39;, &#39;DC_filtered_FC&#39;, &#39;Mono_filtered_FC&#39;, &#39;NK_filtered_FC&#39;, &#39;Platelet_filtered_FC&#39;
    var: &#39;features&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;Cell type_colors&#39;, &#39;Prediction fast mode_colors&#39;, &#39;Prediction p-val_colors&#39;, &#39;Prediction q_colors&#39;, &#39;Prediction standard mode_colors&#39;, &#39;celltype.l1_colors&#39;, &#39;celltype.l2_colors&#39;, &#39;celltype.l3_colors&#39;, &#39;hvg&#39;, &#39;neighbors&#39;, &#39;signature_based_classification&#39;
    obsm: &#39;X_apca&#39;, &#39;X_aumap&#39;, &#39;X_pca&#39;, &#39;X_spca&#39;, &#39;X_umap&#39;, &#39;X_wnn.umap&#39;
    varm: &#39;PCs&#39;, &#39;SPCA&#39;
    obsp: &#39;distances&#39;
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>rm ./data/copy_atlas.h5ad
</pre></div>
</div>
<p>For standardization purposes and because of <strong>Celltypist strict
requirements</strong>, we rescaled counts from 0 to 10000 for each cell and we
log transformed the resulting values, accordingly with <a class="reference external" href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">Scanpy
tutorial</a>.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_11_0.png"><img alt="../../_images/output_11_0.png" src="../../_images/output_11_0.png" style="width: 463px; height: 371px;" /></a>
</div>
<div class="section" id="test-dataset">
<h2>Test dataset<a class="headerlink" href="#test-dataset" title="Permalink to this headline">ÔÉÅ</a></h2>
<p>We <strong>fitted the trained model of each classifier on</strong>
<a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.datasets.pbmc3k.html">PBMC3K</a>
from Satija et al.¬†2015
<a class="reference external" href="https://www.nature.com/articles/nbt.3192">[4]</a>, and we compared the
performances of classification with CIA ones.</p>
<p>NB: the test dataset was normalized in the very same way of the training
dataset.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc3k</span><span class="o">=</span><span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;data/pbmc3k_classified.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="cluster-based-annotation">
<h2>Cluster-based annotation<a class="headerlink" href="#cluster-based-annotation" title="Permalink to this headline">ÔÉÅ</a></h2>
<p>First of all, we <strong>challanged cluster-dependent classifiers</strong>, which
requires that the test dataset must be already divided into meaningful
clusters by setting the proper resolution.</p>
<div class="section" id="scorect">
<h3>Scorect<a class="headerlink" href="#scorect" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To read gmt file  (same signatures used for CIA)</span>
<span class="n">ref_marker</span> <span class="o">=</span> <span class="n">ct</span><span class="o">.</span><span class="n">read_markers_from_file</span><span class="p">(</span><span class="s1">&#39;data/atlas.gmt&#39;</span><span class="p">)</span>

<span class="c1"># To perform differential expression analysis and obtain DEGs</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">var</span><span class="p">),</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">marker_df</span> <span class="o">=</span> <span class="n">ct</span><span class="o">.</span><span class="n">wrangle_ranks_from_anndata</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">)</span>


<span class="c1"># To set the background genes - here, all the genes used to run the differential gene expression test</span>
<span class="n">background</span> <span class="o">=</span> <span class="n">pbmc3k</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

<span class="c1"># to score cell types for each cluster</span>
<span class="n">ct_pval</span><span class="p">,</span> <span class="n">ct_score</span> <span class="o">=</span> <span class="n">ct</span><span class="o">.</span><span class="n">celltype_scores</span><span class="p">(</span><span class="n">nb_bins</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                                        <span class="n">ranked_genes</span><span class="o">=</span><span class="n">marker_df</span><span class="p">,</span>
                                        <span class="n">K_top</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
                                        <span class="n">marker_ref</span><span class="o">=</span><span class="n">ref_marker</span><span class="p">,</span>
                                        <span class="n">background_genes</span><span class="o">=</span><span class="n">background</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">WARNING</span><span class="p">:</span> <span class="n">Default</span> <span class="n">of</span> <span class="n">the</span> <span class="n">method</span> <span class="n">has</span> <span class="n">been</span> <span class="n">changed</span> <span class="n">to</span> <span class="s1">&#39;t-test&#39;</span> <span class="kn">from</span> <span class="s1">&#39;t-test_overestim_var&#39;</span>
<span class="n">Wrangling</span><span class="p">:</span> <span class="n">Number</span> <span class="n">of</span> <span class="n">markers</span> <span class="n">used</span> <span class="ow">in</span> <span class="n">ranked_gene_groups</span><span class="p">:</span>  <span class="mi">13714</span>
<span class="n">Wrangling</span><span class="p">:</span> <span class="n">Groups</span> <span class="n">used</span> <span class="k">for</span> <span class="n">ranking</span><span class="p">:</span> <span class="n">Cell</span> <span class="nb">type</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To assign identity to each cluster depending on the scores</span>
<span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Prediction scorect&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ct</span><span class="o">.</span><span class="n">assign_celltypes</span><span class="p">(</span><span class="n">cluster_assignment</span><span class="o">=</span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">],</span> <span class="n">ct_pval_df</span><span class="o">=</span><span class="n">ct_pval</span><span class="p">,</span> <span class="n">ct_score_df</span><span class="o">=</span><span class="n">ct_score</span><span class="p">,</span> <span class="n">cutoff</span><span class="o">=</span><span class="mf">0.1</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction p-val&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction scorect&#39;</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.4</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">...</span> <span class="n">storing</span> <span class="s1">&#39;Prediction scorect&#39;</span> <span class="k">as</span> <span class="n">categorical</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_20_1.png"><img alt="../../_images/output_20_1.png" src="../../_images/output_20_1.png" style="width: 1907px; height: 375px;" /></a>
</div>
<div class="section" id="performances">
<h3>Performances<a class="headerlink" href="#performances" title="Permalink to this headline">ÔÉÅ</a></h3>
<p>We next exploited <strong>classification_metrics</strong> function of <strong>CIA report
module</strong> to evaluate the performance of each classifier using Satija et
al.¬†2015 <a class="reference external" href="https://www.nature.com/articles/nbt.3192">[4]</a> annotation
as ground truth.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">report</span><span class="o">.</span><span class="n">classification_metrics</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">classification_obs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction p-val&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction scorect&#39;</span><span class="p">],</span> <span class="n">groups_obs</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SE</th>
      <th>SP</th>
      <th>PR</th>
      <th>ACC</th>
      <th>F1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Prediction fast mode</th>
      <td>0.895375</td>
      <td>0.982563</td>
      <td>0.895375</td>
      <td>0.970107</td>
      <td>0.895375</td>
    </tr>
    <tr>
      <th>Prediction p-val</th>
      <td>0.847991</td>
      <td>0.980035</td>
      <td>0.876224</td>
      <td>0.961172</td>
      <td>0.861876</td>
    </tr>
    <tr>
      <th>Prediction scorect</th>
      <td>0.866187</td>
      <td>0.977698</td>
      <td>0.866187</td>
      <td>0.961768</td>
      <td>0.866187</td>
    </tr>
  </tbody>
</table>
</div><p>Despite the good performances of Scorect (F1-score &gt; 0.85, with overall
performances comparable with CIA predictions), 2 whole clusters were
misclassified, meaning that <strong>2 cell populations were not detected</strong>.</p>
</div>
</div>
<div class="section" id="cell-level-classification">
<h2>Cell level classification<a class="headerlink" href="#cell-level-classification" title="Permalink to this headline">ÔÉÅ</a></h2>
<p>Since, like CIA, <strong>Celltypist and Besca</strong> allow to <strong>automatically
classify datasets cell-by-cell</strong>, without requiring the clustering, we
challanged both classifiers at single cell level.</p>
<div class="section" id="celltypist">
<h3>Celltypist<a class="headerlink" href="#celltypist" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To train Celltypist with PBMC atlas feature and ground truth labels (&#39;Cell type&#39; obs).</span>

<span class="c1"># COMPUTATIONALLY HEAVY! results have been stored in &#39;data/model_atlas.pkl&#39;.</span>
<span class="c1">#new_model = celltypist.train(adata, labels = &#39;Cell type&#39;, n_jobs = 32, feature_selection = True)</span>

<span class="c1"># To load trained model</span>
<span class="n">new_model</span><span class="o">=</span><span class="n">celltypist</span><span class="o">.</span><span class="n">Model</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;data/model_atlas_no_other_t.pkl&#39;</span><span class="p">)</span>
<span class="n">new_model</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CellTypist</span> <span class="n">model</span> <span class="k">with</span> <span class="mi">7</span> <span class="n">cell</span> <span class="n">types</span> <span class="ow">and</span> <span class="mi">1280</span> <span class="n">features</span>
    <span class="n">date</span><span class="p">:</span> <span class="mi">2022</span><span class="o">-</span><span class="mi">05</span><span class="o">-</span><span class="mi">26</span> <span class="mi">13</span><span class="p">:</span><span class="mi">54</span><span class="p">:</span><span class="mf">42.584144</span>
    <span class="n">cell</span> <span class="n">types</span><span class="p">:</span> <span class="n">B</span><span class="p">,</span> <span class="n">CD4</span> <span class="n">T</span><span class="p">,</span> <span class="o">...</span><span class="p">,</span> <span class="n">Platelet</span>
    <span class="n">features</span><span class="p">:</span> <span class="n">TTLL10</span><span class="p">,</span> <span class="n">TNFRSF18</span><span class="p">,</span> <span class="o">...</span><span class="p">,</span> <span class="n">MTRNR2L8</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To fit the model and predict pbmc3k cell identities.</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">celltypist</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">to_adata</span><span class="p">(),</span> <span class="n">model</span> <span class="o">=</span> <span class="n">new_model</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>üî¨ Input data has 2638 cells and 13714 genes
üîó Matching reference genes in the model
üß¨ 1157 features used for prediction
‚öñÔ∏è Scaling input data
üñãÔ∏è Predicting labels
‚úÖ Prediction done!
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Prediction celltypist&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">predictions</span><span class="o">.</span><span class="n">predicted_labels</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction p-val&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction celltypist&#39;</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">...</span> <span class="n">storing</span> <span class="s1">&#39;Prediction celltypist&#39;</span> <span class="k">as</span> <span class="n">categorical</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_31_1.png"><img alt="../../_images/output_31_1.png" src="../../_images/output_31_1.png" style="width: 2034px; height: 375px;" /></a>
</div>
<div class="section" id="besca">
<h3>Besca<a class="headerlink" href="#besca" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># to train Besca Logistic Regression model and perform feature selection</span>
<span class="n">adata_train</span><span class="p">,</span> <span class="n">adata_test_corrected</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">merge_data</span><span class="p">([</span><span class="n">adata</span><span class="p">],</span> <span class="n">pbmc3k</span><span class="p">,</span> <span class="n">genes_to_use</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">merge</span> <span class="o">=</span> <span class="s1">&#39;scanorama&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">merging</span> <span class="k">with</span> <span class="n">scanorama</span>
<span class="n">using</span> <span class="n">scanorama</span> <span class="n">rn</span>
<span class="n">Found</span> <span class="mi">350</span> <span class="n">genes</span> <span class="n">among</span> <span class="nb">all</span> <span class="n">datasets</span>
<span class="p">[[</span><span class="mf">0.</span>         <span class="mf">0.94427597</span><span class="p">]</span>
 <span class="p">[</span><span class="mf">0.</span>         <span class="mf">0.</span>        <span class="p">]]</span>
<span class="n">Processing</span> <span class="n">datasets</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">integrating</span> <span class="n">training</span> <span class="nb">set</span>
<span class="n">calculating</span> <span class="n">intersection</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">classifier</span><span class="p">,</span> <span class="n">scaler</span> <span class="o">=</span> <span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">adata_train</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;logistic_regression&#39;</span><span class="p">,</span> <span class="n">celltype</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span> <span class="n">njobs</span><span class="o">=</span><span class="mi">32</span><span class="p">)</span>
<span class="n">bc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">auto_annot</span><span class="o">.</span><span class="n">adata_predict</span><span class="p">(</span><span class="n">classifier</span> <span class="o">=</span> <span class="n">classifier</span><span class="p">,</span> <span class="n">adata_pred</span> <span class="o">=</span> <span class="n">adata_test_corrected</span><span class="p">,</span> <span class="n">adata_orig</span><span class="o">=</span><span class="n">pbmc3k</span><span class="p">,</span>
                               <span class="n">threshold</span> <span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">scaler</span><span class="o">=</span><span class="n">scaler</span><span class="p">)</span>
<span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Prediction besca LR&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;auto_annot&#39;</span><span class="p">]</span>
<span class="k">del</span> <span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;auto_annot&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="mi">32</span><span class="p">)]:</span> <span class="n">Using</span> <span class="n">backend</span> <span class="n">LokyBackend</span> <span class="k">with</span> <span class="mi">32</span> <span class="n">concurrent</span> <span class="n">workers</span><span class="o">.</span>
<span class="p">[</span><span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="mi">32</span><span class="p">)]:</span> <span class="n">Done</span>   <span class="mi">2</span> <span class="n">out</span> <span class="n">of</span>   <span class="mi">5</span> <span class="o">|</span> <span class="n">elapsed</span><span class="p">:</span>  <span class="mf">2.6</span><span class="nb">min</span> <span class="n">remaining</span><span class="p">:</span>  <span class="mf">4.0</span><span class="nb">min</span>
<span class="p">[</span><span class="n">Parallel</span><span class="p">(</span><span class="n">n_jobs</span><span class="o">=</span><span class="mi">32</span><span class="p">)]:</span> <span class="n">Done</span>   <span class="mi">5</span> <span class="n">out</span> <span class="n">of</span>   <span class="mi">5</span> <span class="o">|</span> <span class="n">elapsed</span><span class="p">:</span>  <span class="mf">2.9</span><span class="nb">min</span> <span class="n">finished</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># to train Besca SVM model and perform feature selection</span>

<span class="c1"># COMPUTATIONALLY HEAVY, results of this chunk have been already saved in pbmc3k.obs[&#39;auto_annot&#39;]</span>

<span class="c1"># classifier, scaler = bc.tl.auto_annot.fit(adata_train, method=&#39;linear&#39;, celltype=&#39;Cell type&#39;, njobs=10)</span>
<span class="c1"># bc.tl.auto_annot.adata_predict(classifier = classifier, adata_pred = adata_test_corrected, adata_orig=pbmc3k,</span>
<span class="c1">#                                threshold =0.1, scaler=scaler)</span>
<span class="c1"># pbmc3k.obs[&#39;Prediction besca SVM&#39;]=pbmc3k.obs[&#39;auto_annot&#39;]</span>
<span class="c1"># del pbmc3k.obs[&#39;auto_annot&#39;]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction p-val&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction besca LR&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca SVM&#39;</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">...</span> <span class="n">storing</span> <span class="s1">&#39;Prediction besca LR&#39;</span> <span class="k">as</span> <span class="n">categorical</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_36_1.png"><img alt="../../_images/output_36_1.png" src="../../_images/output_36_1.png" style="width: 2034px; height: 723px;" /></a>
</div>
<div class="section" id="id1">
<h3>Performances<a class="headerlink" href="#id1" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">report</span><span class="o">.</span><span class="n">classification_metrics</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">classification_obs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span>
                                                         <span class="s1">&#39;Prediction p-val&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction celltypist&#39;</span><span class="p">,</span>
                                                         <span class="s1">&#39;Prediction besca LR&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca SVM&#39;</span><span class="p">],</span> <span class="n">groups_obs</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SE</th>
      <th>SP</th>
      <th>PR</th>
      <th>ACC</th>
      <th>F1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Prediction fast mode</th>
      <td>0.895375</td>
      <td>0.982563</td>
      <td>0.895375</td>
      <td>0.970107</td>
      <td>0.895375</td>
    </tr>
    <tr>
      <th>Prediction p-val</th>
      <td>0.847991</td>
      <td>0.980035</td>
      <td>0.876224</td>
      <td>0.961172</td>
      <td>0.861876</td>
    </tr>
    <tr>
      <th>Prediction celltypist</th>
      <td>0.775588</td>
      <td>0.962598</td>
      <td>0.775588</td>
      <td>0.935882</td>
      <td>0.775588</td>
    </tr>
    <tr>
      <th>Prediction besca LR</th>
      <td>0.845337</td>
      <td>0.974223</td>
      <td>0.845337</td>
      <td>0.955811</td>
      <td>0.845337</td>
    </tr>
    <tr>
      <th>Prediction besca SVM</th>
      <td>0.868840</td>
      <td>0.978140</td>
      <td>0.868840</td>
      <td>0.962526</td>
      <td>0.868840</td>
    </tr>
  </tbody>
</table>
</div><p>Notably, <strong>CIA fast mode prediction was the best one</strong>, followed by the
computationally heavier Besca SVM prediction and CIA p-val mode (which
are comparable).</p>
</div>
</div>
<div class="section" id="over-clustering-driven-approach">
<h2>Over-clustering driven approach<a class="headerlink" href="#over-clustering-driven-approach" title="Permalink to this headline">ÔÉÅ</a></h2>
<p><strong>Celltypist</strong> has an interesting feature called <strong>‚Äòmajority voting‚Äô</strong>,
which consist of the refinement of cell identities within local
subclusters after an <strong>over-clustering step</strong>. Basically, <strong>for each
subcluster</strong>, the <strong>label of the most abundant cell type is extended</strong>
to the whole cell group. This is an <strong>additional step that goes beyond
the CIA workflow</strong>, but since we wanted to compare the results at the
best possible conditions of classification, we wrote
<strong>celltypist_majority_vote</strong> function (<strong>external module</strong>) to reproduce
the ‚Äòmajority voting‚Äô approach.</p>
<p>To be more clear, <strong>each of the cell-level classifications were matched
with those 68 clusters</strong> (small groups of very similar cells) and at
each mini-cluster was assigned the most abundat cell type label.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;leiden_5&#39;</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_43_0.png"><img alt="../../_images/output_43_0.png" src="../../_images/output_43_0.png" style="width: 567px; height: 622px;" /></a>
<div class="section" id="celltypist-majority-voting">
<h3>Celltypist majority voting<a class="headerlink" href="#celltypist-majority-voting" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To run Celltypist with majority voting</span>
<span class="n">predictions</span> <span class="o">=</span> <span class="n">celltypist</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">raw</span><span class="o">.</span><span class="n">to_adata</span><span class="p">(),</span> <span class="n">model</span> <span class="o">=</span> <span class="n">new_model</span><span class="p">,</span> <span class="n">majority_voting</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>üî¨ Input data has 2638 cells and 13714 genes
üîó Matching reference genes in the model
üß¨ 1157 features used for prediction
‚öñÔ∏è Scaling input data
üñãÔ∏è Predicting labels
‚úÖ Prediction done!
üëÄ Can not detect a neighborhood graph, will construct one before the over-clustering
‚õìÔ∏è Over-clustering input data with resolution set to 5
üó≥Ô∏è Majority voting the predictions
‚úÖ Majority voting done!
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pbmc3k</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Prediction celltypist majority voting&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">predictions</span><span class="o">.</span><span class="n">predicted_labels</span><span class="p">[</span><span class="s1">&#39;majority_voting&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction celltypist majority voting&#39;</span><span class="p">],</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_47_0.png"><img alt="../../_images/output_47_0.png" src="../../_images/output_47_0.png" style="width: 1134px; height: 375px;" /></a>
</div>
<div class="section" id="majority-voting-embedded-in-cia">
<h3>Majority voting embedded in CIA<a class="headerlink" href="#majority-voting-embedded-in-cia" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># To replicate the majority voting step on CIA annotations exploiting celltypist_majority_vote.</span>
<span class="n">colnames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction p-val&#39;</span><span class="p">]</span>
<span class="n">external</span><span class="o">.</span><span class="n">celltypist_majority_vote</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span><span class="n">classification_obs</span><span class="o">=</span><span class="n">colnames</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Reference</span> <span class="n">annotation</span> <span class="ow">not</span> <span class="n">selected</span><span class="o">.</span>
<span class="n">Computing</span> <span class="n">over</span><span class="o">-</span><span class="n">clustering</span> <span class="k">with</span> <span class="n">leiden</span> <span class="n">algorithm</span> <span class="p">(</span><span class="n">resolution</span><span class="o">=</span> <span class="mi">5</span><span class="p">)</span> <span class="o">...</span>
<span class="n">Dataset</span> <span class="n">has</span> <span class="n">been</span> <span class="n">divided</span> <span class="n">into</span> <span class="mi">69</span> <span class="n">groups</span> <span class="n">accordingly</span> <span class="k">with</span> <span class="n">trascriptional</span> <span class="n">similarities</span><span class="o">.</span>

<span class="n">Over</span><span class="o">-</span><span class="n">clustering</span> <span class="n">result</span> <span class="n">saved</span> <span class="ow">in</span> <span class="n">AnnData</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden_5&quot;</span><span class="p">]</span><span class="o">.</span>
<span class="n">Extending</span> <span class="n">the</span> <span class="n">more</span> <span class="n">represented</span> <span class="n">cell</span> <span class="nb">type</span> <span class="n">label</span> <span class="n">to</span> <span class="n">each</span> <span class="n">cell</span> <span class="n">group</span><span class="o">...</span>

<span class="n">New</span> <span class="n">classification</span> <span class="n">labels</span> <span class="n">have</span> <span class="n">been</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">AnnData</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Prediction fast mode majority voting&quot;</span><span class="p">]</span><span class="o">.</span>

<span class="n">New</span> <span class="n">classification</span> <span class="n">labels</span> <span class="n">have</span> <span class="n">been</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">AnnData</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Prediction p-val majority voting&quot;</span><span class="p">]</span><span class="o">.</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">colnames_mv</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction fast mode majority voting&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction p-val majority voting&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colnames_mv</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_51_0.png"><img alt="../../_images/output_51_0.png" src="../../_images/output_51_0.png" style="width: 1648px; height: 375px;" /></a>
</div>
<div class="section" id="majority-voting-applied-to-besca">
<h3>Majority voting applied to Besca<a class="headerlink" href="#majority-voting-applied-to-besca" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">colnames</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Prediction besca LR&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca SVM&#39;</span><span class="p">]</span>
<span class="n">external</span><span class="o">.</span><span class="n">celltypist_majority_vote</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span><span class="n">classification_obs</span><span class="o">=</span><span class="n">colnames</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Reference</span> <span class="n">annotation</span> <span class="ow">not</span> <span class="n">selected</span><span class="o">.</span>
<span class="n">Computing</span> <span class="n">over</span><span class="o">-</span><span class="n">clustering</span> <span class="k">with</span> <span class="n">leiden</span> <span class="n">algorithm</span> <span class="p">(</span><span class="n">resolution</span><span class="o">=</span> <span class="mi">5</span><span class="p">)</span> <span class="o">...</span>
<span class="n">Dataset</span> <span class="n">has</span> <span class="n">been</span> <span class="n">divided</span> <span class="n">into</span> <span class="mi">69</span> <span class="n">groups</span> <span class="n">accordingly</span> <span class="k">with</span> <span class="n">trascriptional</span> <span class="n">similarities</span><span class="o">.</span>

<span class="n">Over</span><span class="o">-</span><span class="n">clustering</span> <span class="n">result</span> <span class="n">saved</span> <span class="ow">in</span> <span class="n">AnnData</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;leiden_5&quot;</span><span class="p">]</span><span class="o">.</span>
<span class="n">Extending</span> <span class="n">the</span> <span class="n">more</span> <span class="n">represented</span> <span class="n">cell</span> <span class="nb">type</span> <span class="n">label</span> <span class="n">to</span> <span class="n">each</span> <span class="n">cell</span> <span class="n">group</span><span class="o">...</span>

<span class="n">New</span> <span class="n">classification</span> <span class="n">labels</span> <span class="n">have</span> <span class="n">been</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">AnnData</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Prediction besca LR majority voting&quot;</span><span class="p">]</span><span class="o">.</span>

<span class="n">New</span> <span class="n">classification</span> <span class="n">labels</span> <span class="n">have</span> <span class="n">been</span> <span class="n">stored</span> <span class="ow">in</span> <span class="n">AnnData</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Prediction besca SVM majority voting&quot;</span><span class="p">]</span><span class="o">.</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">colnames_mv</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Cell type&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca LR majority voting&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction besca SVM majority voting&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">colnames_mv</span><span class="p">,</span> <span class="n">wspace</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/output_55_0.png"><img alt="../../_images/output_55_0.png" src="../../_images/output_55_0.png" style="width: 1648px; height: 375px;" /></a>
</div>
<div class="section" id="id2">
<h3>Performances<a class="headerlink" href="#id2" title="Permalink to this headline">ÔÉÅ</a></h3>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">report</span><span class="o">.</span><span class="n">classification_metrics</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">classification_obs</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Prediction celltypist majority voting&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction fast mode majority voting&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction p-val majority voting&#39;</span><span class="p">,</span>
             <span class="s1">&#39;Prediction besca LR majority voting&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca SVM majority voting&#39;</span><span class="p">],</span> <span class="n">groups_obs</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SE</th>
      <th>SP</th>
      <th>PR</th>
      <th>ACC</th>
      <th>F1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Prediction celltypist majority voting</th>
      <td>0.921531</td>
      <td>0.986922</td>
      <td>0.921531</td>
      <td>0.977580</td>
      <td>0.921531</td>
    </tr>
    <tr>
      <th>Prediction fast mode majority voting</th>
      <td>0.966262</td>
      <td>0.994377</td>
      <td>0.966262</td>
      <td>0.990361</td>
      <td>0.966262</td>
    </tr>
    <tr>
      <th>Prediction p-val majority voting</th>
      <td>0.966262</td>
      <td>0.994377</td>
      <td>0.966262</td>
      <td>0.990361</td>
      <td>0.966262</td>
    </tr>
    <tr>
      <th>Prediction besca LR majority voting</th>
      <td>0.939348</td>
      <td>0.989891</td>
      <td>0.939348</td>
      <td>0.982671</td>
      <td>0.939348</td>
    </tr>
    <tr>
      <th>Prediction besca SVM majority voting</th>
      <td>0.957544</td>
      <td>0.992924</td>
      <td>0.957544</td>
      <td>0.987870</td>
      <td>0.957544</td>
    </tr>
  </tbody>
</table>
</div><p>Notably, <strong>all the classifications</strong> resulted to have a <strong>very high
F1-score</strong>. However, <strong>CIA performed the best classifications</strong> also
after the majoity voting step, followed by Besca SVM (comparable).</p>
</div>
</div>
<div class="section" id="conclusions">
<h2>Conclusions<a class="headerlink" href="#conclusions" title="Permalink to this headline">ÔÉÅ</a></h2>
<p>In this notebook we challenged the <strong>newest classifiers compatible with
Scanpy</strong> in annotating the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.datasets.pbmc3k.html">PBMC3K
dataset</a>
and compared their perfomances with CIA ones. Here are summarized all
the results:</p>
<ul class="simple">
<li><p>With <strong>cluster-driven approach</strong> (Scorect, at default setting) <strong>it
was impossible to detect 2 cell populations</strong>, since for 2 clusters
the top scored signature wasn‚Äôt the right one. This result
highlighted that if the classification is cluster-dependent eventual
misclassifications can be propagated to entire clusters.</p></li>
<li><p>Overall the <strong>classifiers allowing cell-level annotation were able to
correctly predict the identity of the majority of cells</strong>, indicating
that is possible to accurately annotate the datasets without the
arbitrariety which characterizes the clustering-driven manual
annotation. Among them <strong>CIA resulted to be the best at cell-level
annotation</strong>.</p></li>
<li><p><strong>The winning approach</strong> of this challenge <strong>is cell-level automatic
annotation followed by</strong> over-clustering and <strong>majority voting</strong>.
With this approach, clustering with very high resolution is performed
only after cell-level classification. <strong>When clustering is used to
refine</strong> an already accurate automatic labelling (and so less
influenced by analysts arbitrary decisions), instead of completely
drive the annotation, <strong>results to be an effective way to integrate</strong>
the contributions of <strong>marker genes expression with</strong> the more
general concept of <strong>transcriptional similarity</strong> in defining cell
identity. Also with this approach <strong>CIA obtained the best results</strong>.</p></li>
</ul>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Prediction fast mode&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction p-val&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction celltypist&#39;</span><span class="p">,</span> <span class="s1">&#39;Prediction besca LR&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca SVM&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Prediction fast mode majority voting&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction p-val majority voting&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction celltypist majority voting&#39;</span><span class="p">,</span>
        <span class="s1">&#39;Prediction besca LR majority voting&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction besca SVM majority voting&#39;</span><span class="p">,</span><span class="s1">&#39;Prediction scorect&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">report</span><span class="o">.</span><span class="n">classification_metrics</span><span class="p">(</span><span class="n">pbmc3k</span><span class="p">,</span> <span class="n">classification_obs</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span> <span class="n">groups_obs</span><span class="o">=</span><span class="s1">&#39;Cell type&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>SE</th>
      <th>SP</th>
      <th>PR</th>
      <th>ACC</th>
      <th>F1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Prediction fast mode</th>
      <td>0.895375</td>
      <td>0.982563</td>
      <td>0.895375</td>
      <td>0.970107</td>
      <td>0.895375</td>
    </tr>
    <tr>
      <th>Prediction p-val</th>
      <td>0.847991</td>
      <td>0.980035</td>
      <td>0.876224</td>
      <td>0.961172</td>
      <td>0.861876</td>
    </tr>
    <tr>
      <th>Prediction celltypist</th>
      <td>0.775588</td>
      <td>0.962598</td>
      <td>0.775588</td>
      <td>0.935882</td>
      <td>0.775588</td>
    </tr>
    <tr>
      <th>Prediction besca LR</th>
      <td>0.845337</td>
      <td>0.974223</td>
      <td>0.845337</td>
      <td>0.955811</td>
      <td>0.845337</td>
    </tr>
    <tr>
      <th>Prediction besca SVM</th>
      <td>0.868840</td>
      <td>0.978140</td>
      <td>0.868840</td>
      <td>0.962526</td>
      <td>0.868840</td>
    </tr>
    <tr>
      <th>Prediction fast mode majority voting</th>
      <td>0.966262</td>
      <td>0.994377</td>
      <td>0.966262</td>
      <td>0.990361</td>
      <td>0.966262</td>
    </tr>
    <tr>
      <th>Prediction p-val majority voting</th>
      <td>0.966262</td>
      <td>0.994377</td>
      <td>0.966262</td>
      <td>0.990361</td>
      <td>0.966262</td>
    </tr>
    <tr>
      <th>Prediction celltypist majority voting</th>
      <td>0.921531</td>
      <td>0.986922</td>
      <td>0.921531</td>
      <td>0.977580</td>
      <td>0.921531</td>
    </tr>
    <tr>
      <th>Prediction besca LR majority voting</th>
      <td>0.939348</td>
      <td>0.989891</td>
      <td>0.939348</td>
      <td>0.982671</td>
      <td>0.939348</td>
    </tr>
    <tr>
      <th>Prediction besca SVM majority voting</th>
      <td>0.957544</td>
      <td>0.992924</td>
      <td>0.957544</td>
      <td>0.987870</td>
      <td>0.957544</td>
    </tr>
    <tr>
      <th>Prediction scorect</th>
      <td>0.866187</td>
      <td>0.977698</td>
      <td>0.866187</td>
      <td>0.961768</td>
      <td>0.866187</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../workflow/Cluster_Independent_Annotation.html" class="btn btn-neutral float-left" title="CLUSTER INDEPENDENT ANNOTATION" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../../license.html" class="btn btn-neutral float-right" title="License" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, National Institue of Molecular Genetics (INGM).</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>